tools/infrasys/src/shared.rs

```diff
@@ -0,0 +1,91 @@
+use log::info;
+use rusoto_cloudformation::{CloudFormation, CloudFormationClient, DescribeStacksInput, Parameter};
+use snafu::{ensure, OptionExt, ResultExt};
+use std::{env, thread, time};
+use structopt::StructOpt;
+
+use super::{error, Result};
+
+#[derive(Debug, StructOpt)]
+pub enum KeyRole {
+    Root,
+    Publication,
+}
+
+/// Retrieve a BUILDSYS_* variable that we expect to be set in the environment
+pub fn getenv(var: &str) -> Result<String> {
+    env::var(var).context(error::Environment { var })
+}
+
+/// Generates a parameter type object used to specify parameters in CloudFormation templates
+pub fn get_parameter(key: String, val: String) -> Parameter {
+    Parameter {
+        parameter_key: Some(key),
+        parameter_value: Some(val),
+        ..Default::default()
+    }
+}
+
+/// Polls cfn_client for stack_name in region until it's ready
+/// Once stack is created, we can grab the outputs (before this point, outputs are empty)
+pub async fn get_stack_outputs(
+    cfn_client: &CloudFormationClient,
+    stack_name: &str,
+    region: &str,
+) -> Result<Vec<rusoto_cloudformation::Output>> {
+    let mut stack_outputs = cfn_client
+        .describe_stacks(DescribeStacksInput {
+            stack_name: Some(stack_name.to_string()),
+            ..Default::default()
+        })
+        .await
+        .context(error::DescribeStack { stack_name, region })?
+        .stacks
+        .context(error::ParseResponse {
+            what: "stacks",
+            resource_name: stack_name,
+        })?[0]
+        .clone();
+
+    // Checking that keys have been created so we can return updated outputs
+    let mut status = stack_outputs.stack_status;
+    // Max wait is 30 mins (90s*20s = 1800s = 30mins)
+    let mut max_attempts: u32 = 90;
```

Comment:
30m seems a little excessive, doesn't it?

```suggestion
    // Max wait is 10 mins (30 attempts * 20s = 600s = 10mins)
    let mut max_attempts: u32 = 30;
```

Reply:
I think a 30 minutes timeout is appropriate considering some AWS resources take a long time to stand up in a CFN stack. The goal is to not allow it to run beyond what's reasonable when something gets stuck in CloudFormation. 10 minutes would be too short if we decide to add other resources to stand up in Infrasys (VPCs, etc).

Reply:
KMS keys have been taking quite a while to stand up :sweat_smile:

Reply:
Ok, (sadly) resolving.

Reply:
Although it might be nice to fix the formatting a unit nits.